File: misc/sctp-perf/go/perf_server.go
package main

import (
	"encoding/binary"
	"errors"
	"fmt"
	"net"
	"os"
	"strconv"
	"time"
)

const (
	msgNotification = 0x8000

	frameData   = byte(1)
	frameStop   = byte(2)
	frameResult = byte(3)

	defaultPPID = uint32(0x50524631) // "PRF1"
)

func encodeFrame(kind byte, payload []byte) []byte {
	out := make([]byte, 5+len(payload))
	out[0] = kind
	binary.BigEndian.PutUint32(out[1:5], uint32(len(payload)))
	copy(out[5:], payload)
	return out
}

func decodeFrame(frame []byte) (byte, []byte, error) {
	if len(frame) < 5 {
		return 0, nil, errors.New("short frame")
	}
	size := int(binary.BigEndian.Uint32(frame[1:5]))
	if size != len(frame)-5 {
		return 0, nil, fmt.Errorf("length mismatch: got=%d want=%d", len(frame)-5, size)
	}
	return frame[0], frame[5:], nil
}

func main() {
	host := "127.0.0.1"
	port := 19100
	mode := "rtt"
	iterations := 200
	payloadSize := 256

	if len(os.Args) > 1 {
		host = os.Args[1]
	}
	if len(os.Args) > 2 {
		p, err := strconv.Atoi(os.Args[2])
		if err != nil {
			fmt.Fprintf(os.Stderr, "invalid port: %v\n", err)
			os.Exit(1)
		}
		port = p
	}
	if len(os.Args) > 3 {
		mode = os.Args[3]
	}
	if len(os.Args) > 4 {
		v, err := strconv.Atoi(os.Args[4])
		if err != nil || v <= 0 {
			fmt.Fprintf(os.Stderr, "invalid iterations: %v\n", err)
			os.Exit(1)
		}
		iterations = v
	}
	if len(os.Args) > 5 {
		v, err := strconv.Atoi(os.Args[5])
		if err != nil || v <= 0 {
			fmt.Fprintf(os.Stderr, "invalid payload size: %v\n", err)
			os.Exit(1)
		}
		payloadSize = v
	}
	if mode != "rtt" && mode != "throughput" {
		fmt.Fprintf(os.Stderr, "invalid mode %q (expected rtt|throughput)\n", mode)
		os.Exit(1)
	}

	addr, err := net.ResolveSCTPAddr("sctp4", net.JoinHostPort(host, strconv.Itoa(port)))
	if err != nil {
		fmt.Fprintf(os.Stderr, "ResolveSCTPAddr: %v\n", err)
		os.Exit(1)
	}

	conn, err := net.ListenSCTP("sctp4", addr)
	if err != nil {
		fmt.Fprintf(os.Stderr, "ListenSCTP: %v\n", err)
		os.Exit(1)
	}
	defer conn.Close()

	_ = conn.SetInitOptions(net.SCTPInitOptions{NumOStreams: 8, MaxInStreams: 8})
	_ = conn.SubscribeEvents(net.SCTPEventMask{Association: true, Shutdown: true, DataIO: true})

	fmt.Printf("PERF_SERVER_READY lang=go mode=%s bind=%s iterations=%d size=%d\n", mode, net.JoinHostPort(host, strconv.Itoa(port)), iterations, payloadSize)

	buf := make([]byte, payloadSize+4096)
	var started bool
	var start time.Time
	var msgs int
	var bytesTotal int

	for {
		_ = conn.SetReadDeadline(time.Now().Add(20 * time.Second))
		n, _, flags, peer, info, err := conn.ReadFromSCTP(buf)
		if err != nil {
			fmt.Fprintf(os.Stderr, "ReadFromSCTP: %v\n", err)
			os.Exit(1)
		}
		if flags&msgNotification != 0 {
			continue
		}

		kind, payload, err := decodeFrame(buf[:n])
		if err != nil {
			fmt.Fprintf(os.Stderr, "decode frame: %v\n", err)
			os.Exit(1)
		}

		switch mode {
		case "rtt":
			if kind != frameData {
				fmt.Fprintf(os.Stderr, "unexpected frame type in rtt: %d\n", kind)
				os.Exit(1)
			}
			if !started {
				started = true
				start = time.Now()
			}
			msgs++
			bytesTotal += len(payload)

			snd := &net.SCTPSndInfo{PPID: defaultPPID}
			if info != nil {
				snd.Stream = info.Stream
				snd.PPID = info.PPID
			}
			if _, err := conn.WriteToSCTP(encodeFrame(frameData, payload), peer, snd); err != nil {
				fmt.Fprintf(os.Stderr, "WriteToSCTP echo: %v\n", err)
				os.Exit(1)
			}
			if msgs >= iterations {
				elapsed := time.Since(start).Seconds()
				fmt.Printf("PERF_SERVER_DONE lang=go mode=rtt messages=%d bytes=%d seconds=%.6f\n", msgs, bytesTotal, elapsed)
				return
			}
		case "throughput":
			if kind == frameData {
				if !started {
					started = true
					start = time.Now()
				}
				msgs++
				bytesTotal += len(payload)
				continue
			}
			if kind != frameStop {
				fmt.Fprintf(os.Stderr, "unexpected frame type in throughput: %d\n", kind)
				os.Exit(1)
			}
			elapsed := time.Since(start).Seconds()
			result := []byte(fmt.Sprintf("messages=%d bytes=%d seconds=%.6f", msgs, bytesTotal, elapsed))
			snd := &net.SCTPSndInfo{PPID: defaultPPID}
			if info != nil {
				snd.Stream = info.Stream
				snd.PPID = info.PPID
			}
			if _, err := conn.WriteToSCTP(encodeFrame(frameResult, result), peer, snd); err != nil {
				fmt.Fprintf(os.Stderr, "WriteToSCTP result: %v\n", err)
				os.Exit(1)
			}
			fmt.Printf("PERF_SERVER_DONE lang=go mode=throughput messages=%d bytes=%d seconds=%.6f\n", msgs, bytesTotal, elapsed)
			return
		}
	}
}

File: misc/sctp-perf/rust/src/bin/perf_server.rs
#![feature(sctp)]

use std::env;
use std::io;
use std::net::{SctpListener, SctpSendInfo};
use std::time::{Duration, Instant};

const FRAME_DATA: u8 = 1;
const FRAME_STOP: u8 = 2;
const FRAME_RESULT: u8 = 3;
fn encode_frame(kind: u8, payload: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(5 + payload.len());
    out.push(kind);
    out.extend_from_slice(&(payload.len() as u32).to_be_bytes());
    out.extend_from_slice(payload);
    out
}

fn decode_frame(frame: &[u8]) -> io::Result<(u8, &[u8])> {
    if frame.len() < 5 {
        return Err(io::Error::new(io::ErrorKind::InvalidData, "short frame"));
    }
    let size = u32::from_be_bytes([frame[1], frame[2], frame[3], frame[4]]) as usize;
    if frame.len() != 5 + size {
        return Err(io::Error::new(io::ErrorKind::InvalidData, "frame length mismatch"));
    }
    Ok((frame[0], &frame[5..]))
}

fn parse_args() -> io::Result<(String, u16, String, usize, usize)> {
    let mut host = "127.0.0.1".to_string();
    let mut port: u16 = 19100;
    let mut mode = "rtt".to_string();
    let mut iterations: usize = 200;
    let mut payload_size: usize = 256;

    let args: Vec<String> = env::args().collect();
    if let Some(v) = args.get(1) {
        host = v.clone();
    }
    if let Some(v) = args.get(2) {
        port = v
            .parse::<u16>()
            .map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, format!("invalid port: {e}")))?;
    }
    if let Some(v) = args.get(3) {
        mode = v.clone();
    }
    if let Some(v) = args.get(4) {
        iterations = v.parse::<usize>().map_err(|e| {
            io::Error::new(io::ErrorKind::InvalidInput, format!("invalid iterations: {e}"))
        })?;
    }
    if let Some(v) = args.get(5) {
        payload_size = v.parse::<usize>().map_err(|e| {
            io::Error::new(io::ErrorKind::InvalidInput, format!("invalid payload size: {e}"))
        })?;
    }
    if mode != "rtt" && mode != "throughput" {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            format!("invalid mode {mode} (expected rtt|throughput)"),
        ));
    }
    if iterations == 0 || payload_size == 0 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "iterations and payload size must be positive",
        ));
    }
    Ok((host, port, mode, iterations, payload_size))
}

fn run() -> io::Result<()> {
    let (host, port, mode, iterations, payload_size) = parse_args()?;
    let bind_addr = format!("{host}:{port}");
    let listener = SctpListener::bind(bind_addr.as_str())?;
    let (stream, _peer) = listener.accept()?;
    stream.set_nodelay(true)?;
    stream.set_read_timeout(Some(Duration::from_secs(20)))?;
    stream.set_write_timeout(Some(Duration::from_secs(20)))?;

    println!(
        "PERF_SERVER_READY lang=rust mode={} bind={} iterations={} size={}",
        mode, bind_addr, iterations, payload_size
    );

    let mut recv_buf = vec![0_u8; payload_size + 4096];
    let mut started = false;
    let mut start = Instant::now();
    let mut msgs: usize = 0;
    let mut bytes: usize = 0;

    loop {
        let (n, info) = stream.recv_with_info(&mut recv_buf)?;
        if n == 0 {
            return Err(io::Error::new(io::ErrorKind::UnexpectedEof, "peer closed"));
        }
        let (kind, payload) = decode_frame(&recv_buf[..n])?;

        if mode == "rtt" {
            if kind != FRAME_DATA {
                return Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    format!("unexpected frame type in rtt mode: {kind}"),
                ));
            }
            if !started {
                started = true;
                start = Instant::now();
            }
            msgs += 1;
            bytes += payload.len();

            let send_info = info.map(|i| SctpSendInfo {
                stream: i.stream,
                ppid: i.ppid,
                assoc_id: i.assoc_id,
                ..Default::default()
            });
            stream.send_with_info(&encode_frame(FRAME_DATA, payload), send_info.as_ref())?;
            if msgs >= iterations {
                let elapsed = start.elapsed().as_secs_f64();
                println!(
                    "PERF_SERVER_DONE lang=rust mode=rtt messages={} bytes={} seconds={:.6}",
                    msgs, bytes, elapsed
                );
                return Ok(());
            }
            continue;
        }

        if kind == FRAME_DATA {
            if !started {
                started = true;
                start = Instant::now();
            }
            msgs += 1;
            bytes += payload.len();
            continue;
        }
        if kind != FRAME_STOP {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("unexpected frame type in throughput mode: {kind}"),
            ));
        }
        let elapsed = start.elapsed().as_secs_f64();
        let result = format!("messages={} bytes={} seconds={:.6}", msgs, bytes, elapsed);
        let send_info = info.map(|i| SctpSendInfo {
            stream: i.stream,
            ppid: i.ppid,
            assoc_id: i.assoc_id,
            ..Default::default()
        });
        stream.send_with_info(&encode_frame(FRAME_RESULT, result.as_bytes()), send_info.as_ref())?;
        println!(
            "PERF_SERVER_DONE lang=rust mode=throughput messages={} bytes={} seconds={:.6}",
            msgs, bytes, elapsed
        );
        return Ok(());
    }
}

fn main() {
    if let Err(err) = run() {
        eprintln!("error: {err}");
        std::process::exit(1);
    }
}
